src/satosa/frontends/saml2.py

```diff
@@ -299,12 +300,28 @@ def _handle_authn_response(self, context, internal_response, idp):
             attributes_to_remove = custom_release.get("exclude", [])
             for k in attributes_to_remove:
                 ava.pop(k, None)
-
-        name_id = NameID(text=internal_response.user_id,
-                         format=hash_type_to_saml_name_id_format(
-                             internal_response.user_id_hash_type),
-                         sp_name_qualifier=None,
-                         name_qualifier=None)
+        try:
+            config_nameidformat = self.config['idp_config']['service']['idp']['name_id_format'][0]
+        except KeyError:
+            config_nameidformat = None
+            satosa_logging(logger, logging.DEBUG, "No nameid format configured for frontend idp",
+                           context.state)
+        if config_nameidformat == NAMEID_FORMAT_EMAILADDRESS:
+            nameidfmt = NAMEID_FORMAT_EMAILADDRESS
+            nameidval = internal_response.attributes.get('mail', None)[0]
+            name_id = NameID(text=nameidval,
```

Comment:
`name_id` is going to be set whatever value `config_nameidformat` has. It should be moved out of the `if/else` branches.